src/filter.rs

```diff
@@ -132,12 +132,100 @@ pub(crate) fn unfilter(
         }
     }
 
+    // SIMD Within a Register, see "Unpartitioned Operations With Correction Code"
+    // http://aggregate.org/SWAR/over.html
+    fn swar_add_u32(x: u32, y: u32) -> u32 {
+        // Do 7-bit addition so there's no carry over the most significant bit
+        let n = (x & 0x7f7f7f7f) + (y & 0x7f7f7f7f);
+        // 1-bit parity/XOR addition to fill in the missing MSB
+        n ^ (x ^ y) & 0x80808080
+    }
+
+    fn swar_add_u64(x: u64, y: u64) -> u64 {
+        let n = (x & 0x7f7f7f7f7f7f7f7f) + (y & 0x7f7f7f7f7f7f7f7f);
+        n ^ (x ^ y) & 0x8080808080808080
+    }
+
+    // Average of two integers without overflow or casting to a larger bit-width
+    // http://aggregate.org/MAGIC/#Average%20of%20Integers
+    // The base algorithm:
+    //     (x & y) + ((x ^ y) >> 1)
+    fn swar_avg_u32(x: u32, y: u32) -> u32 {
+        // Calculate 4 averages at once masking out the MSB on the right operand
+        // so right-shifted values don't interfere with their neighbors.
+        (x & y) + (((x ^ y) >> 1) & 0x7f7f7f7f)
+    }
+
+    fn swar_avg_u64(x: u64, y: u64) -> u64 {
+        (x & y) + (((x ^ y) >> 1) & 0x7f7f7f7f7f7f7f7f)
+    }
+
     match filter {
         NoFilter => Ok(()),
         Sub => {
             let current = &mut current[..len];
-            for i in bpp..len {
-                current[i] = current[i].wrapping_add(current[i - bpp]);
+            if bpp > len {
+                return Err("Filtering failed: bytes per pixel is greater than length of row");
+            }
+
+            match tbpp {
+                BytesPerPixel::Three => {
+                    assert!(len > 2);
```

Comment:
I couldn't get the compiler to realize indexing was in bounds without adding these asserts. Information doesn't seem to stick from `tbpp.into_usize()`.